app/src/main/java/eu/kanade/tachiyomi/ui/recent/history/HistoryPresenter.kt

```diff
@@ -40,40 +42,44 @@ class HistoryPresenter(
     private val removeHistoryByMangaId: RemoveHistoryByMangaId = Injekt.get(),
 ) : BasePresenter<HistoryController>() {
 
-    private var _query: MutableStateFlow<String> = MutableStateFlow("")
-    private var _state: MutableStateFlow<HistoryState> = MutableStateFlow(HistoryState.EMPTY)
-    val state: StateFlow<HistoryState> = _state
+    private val _query: MutableStateFlow<String> = MutableStateFlow("")
+    private val _state: MutableStateFlow<HistoryState> = MutableStateFlow(HistoryState.Loading)
+    val state: StateFlow<HistoryState> = _state.asStateFlow()
 
     override fun onCreate(savedState: Bundle?) {
         super.onCreate(savedState)
 
         presenterScope.launchIO {
-            _state.update { state ->
-                state.copy(
-                    list = _query.flatMapLatest { query ->
-                        getHistory.subscribe(query)
-                            .map { pagingData ->
-                                pagingData
-                                    .map {
-                                        UiModel.Item(it)
-                                    }
-                                    .insertSeparators { before, after ->
-                                        val beforeDate = before?.item?.readAt?.time?.toDateKey() ?: Date(0)
-                                        val afterDate = after?.item?.readAt?.time?.toDateKey() ?: Date(0)
-                                        when {
-                                            beforeDate.time != afterDate.time && afterDate.time != 0L -> UiModel.Header(afterDate)
-                                            // Return null to avoid adding a separator between two items.
-                                            else -> null
-                                        }
-                                    }
-                            }
+            _query.collectLatest { query ->
+                getHistory.subscribe(query)
+                    .cachedIn(presenterScope)
+                    .catch { exception ->
+                        _state.emit(HistoryState.Error(exception))
+                    }
+                    .map { pagingData ->
+                        pagingData.toHistoryUiModels()
+                    }.let { uiModelsPagingDataFlow ->
+                        _state.emit(HistoryState.Success(uiModelsPagingDataFlow))
                     }
```

Comment:
`cachedIn` should cache the transformation
```suggestion
            _query.collectLatest { query ->
                getHistory.subscribe(query)
                    .catch { exception ->
                        _state.emit(HistoryState.Error(exception))
                    }
                    .map { pagingData ->
                        pagingData.toHistoryUiModels()
                    }
                    .cachedIn(presenterScope)
                    .let { uiModelsPagingDataFlow ->
                        _state.emit(HistoryState.Success(uiModelsPagingDataFlow))
                    }
```